contracts/utils/MulticallGuard.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.8.0;
+
+contract MulticallGuard {
+    // source: @openzeppelin/contracts/security/ReentrancyGuard.sol (v4.6.2)
+    // Booleans are more expensive than uint256 or any type that takes up a full
+    // word because each write operation emits an extra SLOAD to first read the
+    // slot's contents, replace the bits taken up by the boolean, and then write
+    // back. This is the compiler's defense against contract upgrades and
+    // pointer aliasing, and it cannot be disabled.
+
+    // The values being non-zero value makes deployment a bit more expensive,
+    // but in exchange the refund on every call to nonReentrant will be lower in
+    // amount. Since refunds are capped to a percentage of the total
+    // transaction's gas, it is best to keep them low in cases like this one, to
+    // increase the likelihood of the full refund coming into effect.
+    uint256 private constant _NOT_ENTERED = 1;
+    uint256 private constant _ENTERED = 2;
+
+    uint256 private _multicall_status;
+    /// @dev This is used to prevent reentrancy attacks when a function is called out of a multicall context.
+    uint256 private _function_status;
+    address private _multicall_initial_sender;
+
+    constructor() {
+        _multicall_status = _NOT_ENTERED;
+        _function_status = _NOT_ENTERED;
+        _multicall_initial_sender = address(0);
+    }
+
+    /// @notice Modifier to validate multicall
+    /// @dev This modifier ensures that no surplus native tokens are left in the contract after function execution
+    /// @param noNativeSurplus If true, it checks that no surplus native tokens are left in the contract
+    modifier guardMulticall(bool noNativeSurplus) {
+        _multicall_status = _ENTERED;
```

Comment:
Can add a check here to prevent double calls to this function
```
        // On the first call to guardMulticall, _multicall_status == _NOT_ENTERED
        require(_multicall_status == _NOT_ENTERED, "MulticallGuard: reentrant call");
```